### ISSUE TITLE
Call for Contribution

### ISSUE THREAD

**Issue by kemaswill** [keras-team/keras-contrib#17]
Contributions w.r.t. the following layers and examples are welcomed:

- Layers
   - [Layer Normalization](https://arxiv.org/abs/1607.06450)
   - [PELU](https://arxiv.org/abs/1605.09332) [DONE!]
   - [Batch Renormalization](https://arxiv.org/abs/1702.03275)
- Examples
  - [DenseNets](https://arxiv.org/abs/1608.06993v2) [DONE!]
  - [Residual of Residual Networks](https://arxiv.org/pdf/1608.02908v1.pdf) [DONE!]
  - [Wide ResNets](https://arxiv.org/abs/1605.07146v1)[DONE!]
  - [ResNet in ResNet](https://arxiv.org/abs/1603.08029)

**Comment by titu1994** [comment]
I'd be happy to add PRs, but I need to know what format would suite this
repo.

Generally applications are one offs, they build a specific model, load
weights and then provide the model for the end user.

The models I have are builder types, meaning there are multiple parameter
which must be supplied.

On top of that, the models are not trained on ImageNet. Only Cifar 10 or
100. Also, these weights are for the mid level models, since I don't have a
big enough GPU to train the largest models in the papers.

Any suggestions on these issues ?

Also, while we are discussing this, what about the callback builder for
Snapshot Ensemble? It is achieved via a callback list and not a single
callback.

Should I keep the builder class to simplify the usage, or just put the
callbacks separately and the schedule separately and the user must take the
trouble to set up the model ? Personally I prefer the builder approach.

Somshubra Majumdar

**Comment by kemaswill** [comment]
> The models I have are builder types, meaning there are multiple parameter
which must be supplied.

I think you can provide a builder types, defining several functions with different parameters, such as  def wrn_28_10(): return wrn.create_wide_residual_network(ip, nb_classes=10, N=4, k=10, dropout=0.0, verbose=1) and so on

> On top of that, the models are not trained on ImageNet. Only Cifar 10 or
100.

I think that's fine.

> Also, while we are discussing this, what about the callback builder for
Snapshot Ensemble?

Let's discuss this on the [original issue](https://github.com/farizrahman4u/keras-contrib/issues/10)

**Comment by titu1994** [comment]
@kemaswill Thanks for the clarification. 

Should I add an example script of how to use the DenseNet model as well? As in a script which shows how the DenseNet-40-12 model was trained on the CIFAR-10 dataset?

I'm just about ready to make a PR for DenseNet.

**Comment by kemaswill** [comment]
Yes, go ahead.

**Comment by titu1994** [comment]
@kemaswill I was wondering, isn't BatchNormalization with mode=1 the same as Layer Normalization? Or am I missing something here?

**Comment by titu1994** [comment]
@kemaswill I am attempting to implement the Batch Renormalization paper, but I have a few questions : 

1) How do I get current iteration number in the layer? The paper mentions altering `r_max` and `d_max` after a certain number of iterations gradually, but I don't think Keras layers receive any information about current iteration / epoch number.

2) I would need to replicate a large portion of K.normalize_batch_in_training() because currently it computes and returns `(gamma * ((x - mean) / std)) + beta`, whereas for this algorithm we need is actually  `(((x - mean) / std) * r + d) * gamma + beta`. 

So, should I use K.mean and K.sqrt(K.var(...)) manually? Because I am quite sure that will be a huge performance bottleneck. On the flip side, if I use the theano backend, normalize_batch_in_training() simply calls the theano batch_normalization_train function, which provides high performance, but I can't perform renormalization then. 

Is it worth the performance penalty on theano backend? On tensorflow, I can get the mean and std using tf.nn.moments() but I can't see any similar function for theano.

**Comment by the-moliver** [comment]
@titu1994 
1) I think you should be able to change those during training using callbacks.
2) You could probably use the output of K.normalize_batch_in_training() as such:
```
out = (gamma * ((x - mean) / std)) + beta
new_out = r*out + gamma*d - beta*(r-1)
```
because 
`new_out = (((x - mean) / std) * r + d) * gamma + beta`

In any case K.mean and K.std call the corresponding Theano and Tensorflow functions, so you should always use them if needed.

**Comment by titu1994** [comment]
@the-moliver Wouldn't this be less efficient than simply computing mean and std manually (getting K.mean() and K.var and then compute std)? Because we are computing matrix multiplications of gamma and beta inside normalize_batch_in_training?

**Comment by the-moliver** [comment]
It may actually be more efficient since K.normalize_batch_in_training(), uses specialized theano and tensorflow calls, and returns the mean and variance which can be used to compute d and r. You'll probably have to use something like the trick I mentioned if you want to use the specialized calls.

Yeah I agree it's a little cumbersome, but given that you'd probably want all BatchRenormalization layers to change their parameter together, it should be possible to have a callback change all of them. @bobchennan Did something like this with dropout rates: https://github.com/fchollet/keras/issues/3424
Not sure if there is a better solution...

**Comment by titu1994** [comment]
@the-moliver For now I guess I will keep the standard calls (K.mean and var), but your approach is also valid. 

If I remember correctly, theano and tensorflow have direct cuda calls for batchnorm, so the batchnorm itself would be faster, but then performing more operations on that output may end up costing us. Can't tell without performance tests. 

As for changing the r_max and d_max, that seems like a very hackish way of doing things. Perhaps it is better to just stick with the final r_max and d_max mentioned in the paper (r_max = 3 and d_max = 5). 

According to the paper, practically this leads to problems. Specifically from the paper ; 
```
However, at the beginning of training, when the learning rate was larger, 
it proved important to increase rmax slowly: otherwise, large gradients 
were observed to suddenly and severely increase the loss.
```

**Comment by the-moliver** [comment]
Alternatively, it my be possible to accomplish essentially the same thing by changing one line in the BatchNormalization layer:
```x_normed = K.in_train_phase(x_normed, x_normed_running)```
to
```x_normed = K.in_train_phase((1-param)*x_normed +  param*x_normed_running, x_normed_running)```
with `param` controlling the tradeoff from normalizing with batch statistics to normalizing with running statistics

**Comment by bobchennan** [comment]
```python
class AnnealedDropoutCallback(Callback):
    def __init__(self, N):
        self.N = N
    def on_epoch_end(self, batch, logs={}):
        v      = max(0,1-1.0*batch/self.N)
        #print "epoch",batch,"update",v
        for i in self.model.layers:
            if type(i)==MyDropout:
                K.set_value(i.ratio, v)
```

Should be similar.

**Comment by titu1994** [comment]
@the-moliver That portion of the code if checking if Keras is training or testing. In Mode 0, BN occurs at training time but not at testing time. 

I don't quite understand what your code would do.

**Comment by titu1994** [comment]
@bobchennan I was trying to avoid having a separate callback for just a special class. 

@fchollet Is there any plan in Keras 2 to support such manipulations? I'll hold back on the implementation of this algorithm in that case, until Keras 2 is formalized.

**Comment by the-moliver** [comment]
@titu1994 In the original code `x_normed` and `x_normed_running` are always both computed and then one is chosen for output by `K.in_train_phase` depending on whether its train or test time. My code outputs a weighted combination of `x_normed` and `x_normed_running` during training, and `x_normed_running` at test.

**Comment by titu1994** [comment]
@the-moliver I get that, but what would such weighted combination accomplish? It would be equivalent to the paper.

**Comment by the-moliver** [comment]
@titu1994 Exactly, I was just proposing it as an alternative and simpler implementation.

**Comment by titu1994** [comment]
@the-moliver Oh ok. More important than that is trying to set the value of r_max and d_max.

If it is going to be a part of the Keras 2 spec, we can for the time being use the callback approach and then when Keras 2 is finalized we can drop support for the callback and switch to the native implementation. Is that alright?

**Comment by kemaswill** [comment]
> I was wondering, isn't BatchNormalization with mode=1 the same as Layer Normalization? Or am I missing something here?

@titu1994  BatchNormalization with mode=0,1,2 all use the mean and variance of summed inputs to **a neuron** over a mini-batch of training cases to do normalization, why **layer normalization** use the mean and variance from all of the summed inputs to the neurons in **a layer on a single training case**.

**Comment by titu1994** [comment]
Oh OK. Thanks for the clarification.

**Comment by titu1994** [comment]
@kemaswill @the-moliver So I implemented the algorithm, and to me it looks like it follows the paper exactly in all regards, other than updating the r_max and d_max. 

In a few words, it is performing exceedingly poorly. I am testing it on a WRN-16-4 model and theano backend (works for tensorflow as well). The same model with BatchNormalization performs decently (I am only training for 100 iterations with Adam, not decreasing learning rate or doing full execution of 300 epochs), but the Batch Renormalization model is failing miserably. 
I think a few reasons may be : 

1) The paper mentions explicit updates to running average and mean, and then go on to say the rest is updated via gradient optimization. I am doing that by pulling the update call from the inner if mode == 0 to the general flow of execution. That should be correct right?

2) The loss initially **increases** during the first few iterations and validation accuracy is terrible for these epochs. Is this solely due to the initial r_max and d_max values being so high? To me, it looks like the gradients in the first couple of iterations are too strong due to the high learning rate. 

What could be the possible errors?

**Comment by titu1994** [comment]
@kemaswill @the-moliver So I managed to fix the training problem (i.e. the fact that r_max and d_max needs to be adjusted per iteration). In fact, I had to settle for per epoch updation of the max values, even though the paper suggests specific iterations at which r_max and d_max should be at their max possible value.

How I did it : 
Used the population growth formula (basically a sigmoid formula with extra parameters) 
`N(t) = K / (1 + (K / (N_0) - 1) * e ^ (-(r * t)))` as a function to compute the value at epoch 't'. It has a useful property that one equation can derive two plots, both of which are perfect for smoothly updating the current values of r_max and d_max to their next higher state with different initial values.

The plot below shows the value as t increases. By the 5th-6th epoch, the max value has been reached, and it avoids gradient explosion problem now. Ignore the derivative part.

![biological growth rate](https://cloud.githubusercontent.com/assets/3048602/23098422/762eae70-f612-11e6-9806-cedf346c2bcc.png)

New issues : 

1) Mode 2 Batch Renormalization is learning fast. _Too fast_. I mean to say that with the same WRN model with Batch Renorm and mode 2, I was able to get 67% validation accuracy in the 1st epoch, with _training accuracy at just 51%_. That is in comparison to 35~ % on normal batch normalization with train accuracy of 42%. This may be because CIFAR 10 is too easy to train a network on (WRN-16-4 has 2.8 million parameters in it, but it can't over/underfit in just 1 epoch, I hope).

2) With mode 0, training rises drastically, near 51%, but validation score is very poor at just 21 - 35 % for even the next 5-6 epochs. By this time, training accuracy has already hit 94%, which is weird. This may be due to the formulation of how inference occurs in mode 0 vs mode 2. In mode 0, I am returning the non renornormalized results (simple batch norm results as suggested by the paper) as the output of K.in_training_phase. When I use mode 2, I am sending the renormalized output. 

Perhaps the weights are adjusting to handle just the renormalized values? I am stopping gradient flow from r and d with K.stop_gradients so it can't be the case that the gradients of r and d and affecting the training process.

Any suggestions?

**Comment by titu1994** [comment]
@kemaswill @the-moliver On a final note, should I put Batch Renormalization in Keras directly (branch Keras 2) or should I put this in keras_contrib? Batch renorm is super useful for fast training, especially in case of mode 2 and Batch normalization layers can simply be swapped out for BatchRenormalization layers for an improvement.

@fchollet In your opinion, should this be in keras contrib for now, or merged into Keras 2? I believe that the BRn (Batch Renormalization) layer should not have any conflicts with the changes that will be there in Keras 2. It can be swapped in as a replacement for BN layer pretty much everywhere.

**Comment by gabrieldemarmiesse** [comment]
Closing as outdated, feel free to open another issue with another call for contributions if you want.
